eis_toolkit/raster_processing/reclassify_raster.py

```diff
@@ -0,0 +1,463 @@
+    bands: Sequence[int],
+) -> Tuple[Sequence[np.ndarray], dict]:
+    """Classify raster with quantiles.
+
+    If bands are not given, all bands are used for classification.
+
+    Args:
+        raster: Raster to be classified.
+        number_of_quantiles: The number of quantiles.
+        bands: Selected bands from multiband raster. Indexing begins from one. Defaults to None.
+
+    Returns:
+        Raster classified with quantiles and metadata.
+
+    Raises:
+        InvalidParameterValueException: Bands is not a list, bands is not a list of integers,
+        bands exceeds the number of bands in the raster.
+    """
+
+    out_image = []
+    out_meta = raster.meta.copy()
+
+    bands_to_read = bands if bands is not None else raster.indexes
+    for band in raster.read(bands_to_read):
+
+        _band_exceptions(band)
+
+        numbered_quantiles_band = _raster_with_quantiles(raster, number_of_quantiles)
+
+        out_image.append(numbered_quantiles_band)
+
+    return out_image, out_meta
+
+
+def _raster_with_natural_breaks(  # type: ignore[no-any-unimported]
+    band: np.ndarray,
+    number_of_classes: int,
+) -> Sequence[np.ndarray]:
+
+    out_image = []
+
+    breaks = mc.JenksCaspall(band, number_of_classes)
+    data = np.digitize(band, np.sort(breaks.bins))
+
+    out_image.append(data)
+
+    return out_image
+
+
+@beartype
+def raster_with_natural_breaks(  # type: ignore[no-any-unimported]
+    raster: rasterio.io.DatasetReader,
+    number_of_classes: int,
+    bands: Sequence[int],
+) -> Tuple[Sequence[np.ndarray], dict]:
+    """Classify raster with natural breaks (Jenks Caspall).
+
+    If bands are not given, all bands are used for classification.
+
+    Args:
+        raster: Raster to be classified.
+        number_of_classes: The number of classes.
+        bands: Selected bands from multiband raster. Indexing begins from one. Defaults to None.
+
+    Returns:
+        Raster classified with natural breaks (Jenks Caspall) and metadata.
+
+    Raises:
+        InvalidParameterValueException: Bands is not a list, bands is not a list of integers,
+        bands exceeds the number of bands in the raster.
+    """
+
+    out_image = []
+    out_meta = raster.meta.copy()
+
+    bands_to_read = bands if bands is not None else raster.indexes
+    for band in raster.read(bands_to_read):
+
+        _band_exceptions(band)
+
+        natural_breaks_band = _raster_with_natural_breaks(raster, number_of_classes)
+
+        out_image.append(natural_breaks_band)
+
+    return out_image, out_meta
+
+
+def _raster_with_geometrical_intervals(
+    band: np.ndarray, number_of_classes: int, nan_value: Union[int, float]
+) -> Sequence[np.ndarray]:
+
+    out_image = []
+    is_integer = np.issubdtype(band.dtype, int)
+    is_float = np.issubdtype(band.dtype, float)
+
+    if is_integer:
+        mask = band == nan_value
+        band[mask] = nan_value
+    elif is_float:
+        band[band == nan_value] = np.nan
```

Comment:
Why is the logic different for integers and floats? If the `nan` value for integers is a normal integer (e.g. -9999) it will affect the interval determination, will it not? I think `np.nan` should also be set to integer bands unless I am not understanding things correctly.

Reply:
Ah yes true. In this case I believe you need to use `numpy.ma.masked_array` to make sure the chosen integer nan value does not affect the computations. 

~~~python
# nan_value is either a set integer (e.g. -9999) or np.nan
mask = band == nan_value
masked_array = np.ma.masked_array(data=band, mask=mask)

median_value = np.ma.median(masked_array)
...
~~~

You do not need to check for `np.nan` values anywhere after this i.e. no `nan_condition`. Just use the same name methods from `np.ma` instead of `np`. E.g. `np.where` -> `np.ma.where`. You should check that the output, however, is not a masked array. A masked array can be transformed back to a normal array with `np.array(masked_array)`. This transformation will add back the earlier defined nan_value to the array.